File: src/main/java/io/github/bot/terminal/application/domains/monitoring/rest/MonitoringTasksRestConverter.java
package io.github.bot.terminal.application.domains.monitoring.rest;


import io.github.bot.terminal.application.domains.common.action.ActionRestConverter;
import io.github.bot.terminal.application.domains.common.conditions.ConditionRestConverter;
import io.github.bot.terminal.application.domains.common.notify.NotifyRestConverter;
import io.github.bot.terminal.application.domains.monitoring.repository.MonitoringTaskDetails;
import io.github.bot.terminal.application.domains.monitoring.rest.dto.MonitoringTaskDTO;
import io.github.bot.terminal.application.domains.monitoring.rest.requests.MonitoringTaskRequest;
import io.github.bot.terminal.application.domains.monitoring.values.MonitoringTaskState;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Service;

import java.util.stream.Collectors;


@Service
@RequiredArgsConstructor
public class MonitoringTasksRestConverter {

    private final ActionRestConverter actionRestConverter;
    private final ConditionRestConverter conditionRestConverter;
    private final NotifyRestConverter notifyRestConverter;

    MonitoringTaskDetails mapToDetails(MonitoringTaskRequest request) {
        MonitoringTaskDetails details = new MonitoringTaskDetails();
        details.setName(request.getName());

        if(StringUtils.isNotBlank(request.getState())) {
            details.setState(MonitoringTaskState.valueOf(request.getState()));
        }

        if(request.getAction() != null) {
            details.setActionDetails(actionRestConverter.mapToDetails(request.getAction()));
        }

        details.setCron(request.getCron());
        details.setConditions(request.getConditions().stream()
                .map(c -> conditionRestConverter.mapToDetails(c)).collect(Collectors.toList()));
        details.setNotifyList(request.getNotifyList().stream()
                .map(c -> notifyRestConverter.mapToDetails(c)).collect(Collectors.toList()));
        return details;
    }

    MonitoringTaskDTO mapToDto(MonitoringTaskDetails details) {
        MonitoringTaskDTO dto = new MonitoringTaskDTO();
        dto.setId(details.getId());
        dto.setName(details.getName());
        dto.setState(details.getState().name());
        if(details.getActionDetails() != null) {
            dto.setAction(actionRestConverter.mapToDto(details.getActionDetails()));
        }
        dto.setCron(details.getCron());
        dto.setConditions(details.getConditions().stream()
                .map(c -> conditionRestConverter.mapToDto(c)).collect(Collectors.toList()));
        dto.setNotifyList(details.getNotifyList().stream()
                .map(n -> notifyRestConverter.mapToDto(n)).collect(Collectors.toList()));
        return dto;
    }

}

File: src/main/kotlin/io/github/bot/terminal/application/domains/monitoring/rest/MonitoringTasksRestConverter.kt
package io.github.bot.terminal.application.domains.monitoring.rest

import io.github.bot.terminal.application.domains.common.action.ActionRestConverter
import io.github.bot.terminal.application.domains.common.conditions.ConditionRestConverter
import io.github.bot.terminal.application.domains.common.conditions.entity.ConditionDetails
import io.github.bot.terminal.application.domains.common.conditions.requests.ConditionRequest
import io.github.bot.terminal.application.domains.common.notify.NotifyRestConverter
import io.github.bot.terminal.application.domains.common.notify.entity.NotifyDetails
import io.github.bot.terminal.application.domains.common.notify.requests.NotifyRequest
import io.github.bot.terminal.application.domains.monitoring.repository.MonitoringTaskDetails
import io.github.bot.terminal.application.domains.monitoring.rest.dto.MonitoringTaskDTO
import io.github.bot.terminal.application.domains.monitoring.rest.requests.MonitoringTaskRequest
import org.springframework.stereotype.Service

@Service
class MonitoringTasksRestConverter(
        private val actionRestConverter: ActionRestConverter,
        private val conditionRestConverter: ConditionRestConverter,
        private val notifyRestConverter: NotifyRestConverter
) {


    fun mapToDetails(request: MonitoringTaskRequest): MonitoringTaskDetails {
        val conditions = request.conditions
                .map { c: ConditionRequest -> conditionRestConverter.mapToDetails(c) }
        val notifyList = request.notifyList
                .map { c: NotifyRequest -> notifyRestConverter.mapToDetails(c) }
        return MonitoringTaskDetails(
                name = request.name,
                isEnabled = request.enabled,
                actionDetails = actionRestConverter.mapToDetails(request.action),
                cron = request.cron,
                notifyList = notifyList,
                conditions = conditions
        )
    }

    fun mapToDto(details: MonitoringTaskDetails): MonitoringTaskDTO {
        val conditions = details.conditions
                .map { c: ConditionDetails -> conditionRestConverter.mapToDto(c) }
        val notifyList = details.notifyList
                .map { n: NotifyDetails -> notifyRestConverter.mapToDto(n) }
        return MonitoringTaskDTO(
                id = details.id,
                name = details.name,
                enabled = details.isEnabled,
                action = actionRestConverter.mapToDto(details.actionDetails),
                cron = details.cron,
                notifyList = notifyList,
                conditions = conditions
        )
    }
}